### ISSUE TITLE
Exclude node_modules folders from Spotlight indexing

### ISSUE THREAD

**Issue by as-com** [yarnpkg/yarn#6453]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
node_modules folders are indexed by Spotlight on macOS

**What is the expected behavior?**
node_modules folders are (possibly optionally) excluded from indexing

This can be done by putting a `.metadata_never_index` in node_modules whenever the folder is created.

**Comment by burkeshartsis** [comment]
It seems like this only just became an issue when updating macOS to Mojave 🤷‍♂️

**Comment by as-com** [comment]
@burkeshartsis I thought node_modules polluting spotlight search results and high CPU usage from indexing always been an issue.

**Comment by tcrossland** [comment]
See also https://github.com/npm/npm/issues/15346

**Comment by jrz** [comment]
15346 says the "true" solution is add folder names to the index. This is pure nonsense.  It should be possible to add the metadata_never_index, OR to set a file attribute.
It's either this, or having a global repository and better module loading (and only have spotlight index the files once).

**Comment by chrisbull** [comment]
Found this quick script to add .metadat_never_index to your node_modules folders. 
```
find . -type d -name "node_modules" -exec touch "{}/.metadata_never_index" \;
```

I added it as a bash alias, so I can use apply quickly.
```
alias fix-spotlight='find . -type d -name "node_modules" -exec touch "{}/.metadata_never_index" \;'
```

Just thought I'd share here, for anyone else that is looking. :)

**Comment by kiddyfurby** [comment]
I am sorry but will yarn consider this feature?

**Comment by dicksonhk** [comment]
Based on @chrisbull 's idea, this script does the same thing faster by not going inside any node_modules directories found (coz the whole node_modules directory had already been ignored)
`find -path '*node_modules/*' -prune -o -type d -name 'node_modules' -exec touch '{}/.metadata_never_index' \;`
For myself, I run it on the home directory, so I exclude hidden directories and Library directory to speed things up
`find . -type d -path './.*' -prune -o -path './Pictures*' -prune -o -path './Library*' -prune -o -path '*node_modules/*' -prune -o -type d -name 'node_modules' -exec touch '{}/.metadata_never_index' \; -print`
I also use the -print option so I know which folder is affected

**Comment by justin808** [comment]
For those using @dicksonhk's answer (thanks!) on MacOS, you have to have a dot or other directory after the command "find".

Maybe the same for other linux.

I agree that yarn should automatically add this file, although this is MacOS specific. Maybe some hook is better?

**Comment by dicksonhk** [comment]
> For those using @dicksonhk's answer (thanks!) on MacOS, you have to have a dot or other directory after the command "find".
> 
> Maybe the same for other linux.
> 
> I agree that yarn should automatically add this file, although this is MacOS specific. Maybe some hook is better?

@justin808 You're right! I edited my answer to add ` . -type d ` and it should work fine!
And if you are running the command in the home directory like me you should take a look at the second command which excludes unnecessary directories :)

**Comment by joaomnb** [comment]
My Spotlight seems to ignore these `.metadata_never_index` files, every `node_modules` folder is still indexed... I'm running macOS Mojave. Anyone had the same problem?

**Comment by rolandleth** [comment]
Until this gets added — if it will — another approach would be to add a `postinstall` script: 

```json
"postinstall": "touch ./node_modules/.metadata_never_index"
```

Also, I didn't manage to make @dicksonhk's script to work (maybe because I'm using fish, dunno), but I did find another handy little script: https://github.com/Willian-Zhang/Ignore-node_modules-in-Spotlight.

**Comment by c-bandy** [comment]
> There is no `node_modules` folder at preinstall, though, if you come from a clean slate.

```json
"preinstall": "mkdir node_modules && touch ./node_modules/.metadata_never_index"
```

**Comment by rolandleth** [comment]
Ha, nice call. Sadly, the file gets deleted after `yarn` finishes, no idea why 🤷🏻‍♂️

**Comment by MariuzM** [comment]
Guys does adding .metadata_never_index really work? Because it's still indexing node_modules and subfolders. I've ran 
`sudo find . -type d -path './.*' -prune -o -path './Pictures*' -prune -o -path './Library*' -prune -o -path '*node_modules/*' -prune -o -type d -name 'node_modules' -exec touch '{}/.metadata_never_index' \; -print
`
And then I cleared my index by using 
```
sudo mdutil -i off /
sudo rm -rf /.Spotlight*
sudo mdutil -i on /
sudo mdutil -E /
```

And noticed that it still indexing node folders/subfolders

**Comment by shanedg** [comment]
I didn't see any change with `.metadata_never_index` (caveat that I don't think I cleared my index after adding) but adding my projects folders to to the Spotlight privacy list does seem to have helped. I haven't noticed high resource usage by Spotlight processes since (Mojave and Catalina).

![image](https://user-images.githubusercontent.com/1653199/69011359-9d972100-0937-11ea-9ff6-c08df46534b2.png)

**Comment by joaomnb** [comment]
Same. As I had said before, `.metadata_never_index` does not work on my system. It didn't with macOS Mojave, and still doesn't now with Catalina.

**Comment by tsujp** [comment]
I've read that `.noindex` might be an option for Spotify? Do you see any results with that @joaomnb ?

**Comment by joaomnb** [comment]
No, doesn't work for me either. Spotlight still indexes everything...

**Comment by joaomnb** [comment]
I did some research on that `.noindex` thing and found [this](https://www.macobserver.com/tips/how-to/stop-spotlight-indexing/).

This works in a different way. If you want the folder `test/` to be ignored by Spotlight, you can rename it to `test.noindex/`. While this does work, renaming folders is obviously not an acceptable solution...

**Comment by nemosmithasf** [comment]
.metadata_never_index doesn't work for me either on Catalina

**Comment by Strajk** [comment]
I've glued together solutions found here and around the internets and automated it:

* Get current Spotlight Exclusions
* Find all node_modules dirs in a specified directory
* Add new ones to Spotlight Exclusions
* Restart system service to apply changes

It's definitely not a code that I would be proud of, but it works :) 

https://github.com/Strajk/setup/blob/master/programs/prevent-spotlight-from-indexing-node-modules.js

![image](https://user-images.githubusercontent.com/697301/86780800-6c519980-c05d-11ea-8e33-a2a410393d60.png)

**Comment by wiegell** [comment]
> I've glued together solutions found here and around the internets and automated it:
> 
> * Get current Spotlight Exclusions
> * Find all node_modules dirs in a specified directory
> * Add new ones to Spotlight Exclusions
> * Restart system service to apply changes
> 
> It's definitely not a code that I would be proud of, but it works :)
> 
> https://github.com/Strajk/setup/blob/master/programs/prevent-spotlight-from-indexing-node-modules.js
> 
> ![image](https://user-images.githubusercontent.com/697301/86780800-6c519980-c05d-11ea-8e33-a2a410393d60.png)

Is this actually working for anyone?? Find candidates are not evaluated??

**Comment by Strajk** [comment]
@wiegell still working for me :) can you clarify what do you mean by not being evaluated?

I've written the script intentionally in a very simple, step-by-step style so it's easy to walk-through and debug. I recommend running it in debug mode (I use WebStorm) and checking what each line does :)

**Comment by cblims** [comment]
You are missing execSync on line 21:
```javascript
const candidates = cleanArray((`find ${DIR} -type d -name 'node_modules' -prune\`))
```
This works for me:
```javascript
const candidates = cleanArray(execSync(`find ${DIR} -type d -name 'node_modules' -prune`))
```

**Comment by kuncevic** [comment]
@Strajk just tried to give it to go and got an error
```
Candidates: 11, Existing: 0, To add: 11
child_process.js:669
    throw err;
    ^

Error: Command failed: printf "MYPWD" | sudo -S /usr/libexec/PlistBuddy -c "Add :Exclusions: string /Users/A/Work/Projects/frontendwatch/service/node_modules" /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist

```

I am on the latest Big Sur 11.2.1

**Comment by kuncevic** [comment]
@Strajk just tried to give it to go and got an error
```
Candidates: 11, Existing: 0, To add: 11
child_process.js:669
    throw err;
    ^

Error: Command failed: printf "MYPWD" | sudo -S /usr/libexec/PlistBuddy -c "Add :Exclusions: string /Users/User/Work/frontendwatch/service/node_modules" /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist

```

I am on the latest Big Sur 11.2.1

**Comment by jasonbarry** [comment]
Sorry, it's not clear to me and it hasn't been brought up yet – why not just put all your JS projects into ~/Sites, and drag the Sites directory into the Spotlight privacy pane? Just exclude all your web projects from being Spotlight indexed in general since you can ⌘P in your IDE to find what you want. Then npm/yarn installs won't trigger an mds_stores reindex.

**Comment by kuncevic** [comment]
@jasonbarry 
✅ done 
![image](https://user-images.githubusercontent.com/1283734/107742917-056b3000-6d64-11eb-9999-31efad544ef6.png)

However I wonder how to fix the error above 🤪

**Comment by jeffhuys** [comment]
> Sorry, it's not clear to me and it hasn't been brought up yet – why not just put all your JS projects into ~/Sites, and drag the Sites directory into the Spotlight privacy pane? Just exclude all your web projects from being Spotlight indexed in general since you can ⌘P in your IDE to find what you want. Then npm/yarn installs won't trigger an mds_stores reindex.

You don't know how many times I've been saved by spotlight searching a snippet of code that I KNEW I had somewhere, but didn't know where. Also, webstorm is much more of an energy hog than spotlight is (after indexing, that is).

**Comment by v-kiniv** [comment]
> @Strajk just tried to give it to go and got an error
> 
> ```
> Candidates: 11, Existing: 0, To add: 11
> child_process.js:669
>     throw err;
>     ^
> 
> Error: Command failed: printf "MYPWD" | sudo -S /usr/libexec/PlistBuddy -c "Add :Exclusions: string /Users/User/Work/frontendwatch/service/node_modules" /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist
> ```
> 
> I am on the latest Big Sur 11.2.1
> 
> UPD:
> 
> If I just run this command `sudo -S /usr/libexec/PlistBuddy -c "Add :Exclusions: string /Users/User/Work/frontendwatch/service/node_modules" /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist` getting an error `Error Reading File: /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist`
> 
> Tried to get manually to this folder, seems like no permissions to get in to folder ` /System/Volumes/Data/.Spotlight-V100/`

I had the same issue and resolved it by granting `Full Disk Access` permission to `Terminal.app`(or any other app you're using to run the script):

 
<img width="780" alt="Screen Shot 2021-07-26 at 21 44 17" src="https://user-images.githubusercontent.com/1661016/127042270-cd13c005-c442-4284-9298-6d9a06d2dead.png">

**Comment by JamesLIAOHJ** [comment]
Base on @chrisbull and @dicksonhk ， those find command is still create too mouch file on my macOS Big Sur 11.6. so I modified the command 

`find -E /Volumes/MacData/workspace -type d -name 'node_modules' -not -regex ".+/node_modules/.+/node_modules"  -exec touch '{}/.metadata_never_index' \; -print`

this command affect folders that have no `/node_modules/` in the middle  but `/node_modules` at the end.

**Comment by iloveitaly** [comment]
@JamesLIAOHJ I thought apple removed support for `metadata_never_index` ([ref](https://apple.stackexchange.com/questions/381646/does-metadata-never-index-work-for-keeping-folders-out-of-spotlight-in-catalina)). Are you finding that this magic file works for you?

**Comment by fengerzh** [comment]
No, the only way to stop it is to add your whole project folder into "Privacy of spotlight". Evil Apple.

**Comment by kenandalda** [comment]
I use one of the LLMs to complete the shell script, and it worked!

```shell
#!/bin/bash

# Function to exclude a path
exclude_from_spotlight() {
  local folder=$1
  if [ -d "$folder" ]; then
    echo "Excluding: $folder"
    # sudo mdutil -i off "$folder" > /dev/null 2>&1
    # sudo mdutil -E "$folder" > /dev/null 2>&1
    sudo /usr/libexec/PlistBuddy -c "Add :Exclusions: string $folder" /System/Volumes/Data/.Spotlight-V100/VolumeConfiguration.plist
  else
    echo "Path does not exist or is invalid: $folder"
  fi
}

# Function to find top-level node_modules and .git folders
find_top_level_folders() {
  local folder_name=$1
  local search_path=$2
  find "$search_path" -type d -name "$folder_name" -prune 2>/dev/null
}

# Main logic
if [ -z "$1" ]; then
  echo "Please enter the folder path to search. Example: ./exclude_spotlight.sh /Users/your_username"
  exit 1
fi

search_path=$1

echo "Starting to exclude top-level node_modules and .git folders in path $search_path..."

# Exclude all top-level node_modules folders
find_top_level_folders "node_modules" "$search_path" | while read -r folder; do
  exclude_from_spotlight "$folder"
done

# Exclude all top-level .git folders
find_top_level_folders ".git" "$search_path" | while read -r folder; do
  exclude_from_spotlight "$folder"
done

echo "All specified folders have been excluded from Spotlight."

``` 

Just save the code to `exclude_spotlight.sh`, and run it like:

```shell
./exclude_spotlight.sh /Users/your_username
```

**Comment by jheng-jie** [comment]
What the hell is Apple doing? Now even updating the Exclusions via command line is useless — after a reboot, the system forcibly syncs with whatever is in the Spotlight Privacy settings. Files that should be excluded still show up with mdls, meaning they're still being indexed. The only way to actually stop it is by manually adding folders to the Privacy tab, but there's no CLI to do that. I have dozens of Node projects on my machine, and every single one is getting indexed. It's insanely annoying.

**Comment by roccomuso** [comment]
that's really hitting on the system performance. Decided to just exclude from Spotlight my work directory 👀
